src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -353,21 +354,129 @@ private class ClassWithReference
             public ClassWithExtensionProperty MyReference { get; set; }
         }
 
+        [Theory]
+        [InlineData(@"{""MyIntMissing"":2,""MyReference"":{""MyIntMissingChild"":3}}")]
+        [InlineData(@"{""MyReference"":{""MyIntMissingChild"":3},""MyIntMissing"":2}")]
+        [InlineData(@"{""MyReference"":{""MyNestedClass"":null,""MyInt"":0,""MyIntMissingChild"":3},""MyIntMissing"":2}")]
+        public static void NestedClass(string json)
+        {
+            ClassWithReference obj;
+
+            void Verify()
+            {
+                Assert.IsType<JsonElement>(obj.MyOverflow["MyIntMissing"]);
+                Assert.Equal(1, obj.MyOverflow.Count);
+                Assert.Equal(2, obj.MyOverflow["MyIntMissing"].GetInt32());
+
+                ClassWithExtensionProperty child = obj.MyReference;
+
+                Assert.IsType<JsonElement>(child.MyOverflow["MyIntMissingChild"]);
+                Assert.IsType<JsonElement>(child.MyOverflow["MyIntMissingChild"]);
+                Assert.Equal(1, child.MyOverflow.Count);
+                Assert.Equal(3, child.MyOverflow["MyIntMissingChild"].GetInt32());
+                Assert.Null(child.MyNestedClass);
+                Assert.Equal(0, child.MyInt);
+            }
+
+            obj = JsonSerializer.Deserialize<ClassWithReference>(json);
+            Verify();
+
+            // Round-trip the json and verify.
+            json = JsonSerializer.Serialize(obj);
+            obj = JsonSerializer.Deserialize<ClassWithReference>(json);
+            Verify();
+        }
+
+        private class ParentClassWithObject
+        {
+            public string Text { get; set; }
+            public ChildClassWithObject Child { get; set; }
+
+            [JsonExtensionData]
+            public Dictionary<string, object> ExtensionData { get; set; } = new Dictionary<string, object>();
```

Comment:
Is it worth adding some more test permutations for `JsonExtensionData`, particularly around deeper nesting and order in which the property appears?

So, let's say parent and grand-child contains the extension data.
Also, the property isn't the last one in the POCO, and we have a child property after the extension data within the parent.

Also, regardless of the property order in the POCO, we may want to have the "extra data" appear within the JSON being deserialized in various different orders.

Reply:
Note the order of the properties on the object doesn't matter. Extension data is handled independently and always serialized after the object's properties.

Reply:
What about order of the properties within the JSON payload itself?

Reply:
Yes that matters -- the "NestedClass" test has different ordering that verifies that. Basically if the last property is missing (and a child object) then the issue occurs.

Reply:
> Yes that matters -- the "NestedClass" test has different ordering that verifies that. 

Perfect.

Reply:
Also I manually verified the parent.child.grandchild case with only parent and grandchild having extension data; didn't want to overly-complicate the tests so I didn't add it. Each level is independent and we already have separate tests that verify > 2 levels.